src/lib/projectConfigValidation.ts

```diff
@@ -0,0 +1,439 @@
+      if (error.params.allowedValues) {
+        message = `Try one of: ${error.params.allowedValues.join(', ')}`;
+      } else {
+        message = `Value not allowed. Other errors can cause this`;
+      }
+      break;
+    case AjvErrorKeyword.Maximum:
+    case AjvErrorKeyword.Minimum:
+    case AjvErrorKeyword.ExclusiveMaximum:
+    case AjvErrorKeyword.ExclusiveMinimum:
+      message = `This value must ${error.params.comparison}${error.params.limit}`;
+      break;
+    case AjvErrorKeyword.MaxItems:
+      message = `Value is too long, needs at most ${error.params.limit} items`;
+      break;
+    case AjvErrorKeyword.MaxLength:
+      message = `Value is too long, needs at most ${error.params.limit} characters`;
+      break;
+    case AjvErrorKeyword.MaxProperties:
+      message = `Too many properties, need at most ${error.params.limit}`;
+      break;
+    case AjvErrorKeyword.MinItems:
+      message = `Value is too short, needs at least ${error.params.limit} items`;
+      break;
+    case AjvErrorKeyword.MinLength:
+      message = `Value is too short, needs at least ${error.params.limit} characters`;
+      break;
+    case AjvErrorKeyword.MinProperties:
+      message = `Too few properties, need at least ${error.params.limit}`;
+      break;
+    case AjvErrorKeyword.MultipleOf:
+      message = `This value must be a multiple of ${error.params.multipleOf}`;
+      break;
+    case AjvErrorKeyword.OneOf:
+      onlyKeyNode = true;
+      message = `Does not match any of the allowed schemas`;
+      break;
+    case AjvErrorKeyword.Pattern:
+      message = `Value does not match pattern: ${error.params.pattern}`;
+      break;
+    case AjvErrorKeyword.Required:
+      onlyKeyNode = true;
+      message = getMissingRequiredFieldError(error.params.missingProperty);
+      break;
+    case AjvErrorKeyword.Type:
+      message = `Incorrect type: ${error.message}`;
+      break;
+    default:
+      break;
+  }
+  return { message, nodePath, onlyKeyNode };
+}
+
+function diagnosticDeduplicationKey(
+  nodePath: (string | number)[],
+  keyword: string,
+  params: any
+): string {
+  return `${keyword}::${nodePath.join('/')}::${JSON.stringify(params)}`;
+}
+
+function getKeyNode(node: jsonc.Node): jsonc.Node | undefined {
+  return node.parent?.type === 'property' ? node.parent.children?.[0] : node;
+}
+
+async function getCachedSchemasForAllTypes(
+  projectRoot: Uri,
+  platformVersion: string,
+  accountId: number
+): Promise<SchemaCacheEntry> {
+  // If schema cache has a schema for this version, return it
+  // If there is an error, we're going to return that too
+  if (schemaCache[platformVersion]) {
+    return schemaCache[platformVersion];
+  }
+  let response: Record<string, AnySchema>;
+  try {
+    response = await getIntermediateRepresentationSchema({
+      projectSourceDir: projectRoot.fsPath,
+      platformVersion,
+      accountId,
+    });
+    let byType: Record<string, SchemaAndValidator> = {};
```

Comment:
Nit: could we give this a slightly more descriptive name?  Also, I believe this could be a const since we aren't reassigning it.